---
title: "FIGURE 1 - PATHWAY ENRICHMENT"
author: "C P O'Meara"
geometry: "left=3cm,right=3cm,top=2cm,bottom=2cm"
output:
  pdf_document: default
---
```{r message=FALSE, warning=FALSE, include=FALSE,tidy=TRUE, tidy.opts=list(width.cutoff=60)}
library(readxl)
library(ReactomePA)
library(clusterProfiler)
library(DOSE)
library(dplyr)
library(org.Dr.eg.db)
library(KEGGREST)
library(readr)
library(ggplot2)
library(gplots)
library(reshape2)
library(stringr)
library(knitr)

#IMPORT ANNOTATION FILES
MOUSE.ORTHOLOGS <- read.delim("MOUSE.ORTHOLOGS.txt", stringsAsFactors=FALSE)
NODE_PATHWAY_ANNOTATION <- read.delim("NODE_PATHWAY_ANNOTATION.txt", stringsAsFactors=FALSE)
MUTANT.LINE.LOG2.FC <- read.delim("MUTANT.LINE.LOG2.FC.txt", stringsAsFactors=FALSE)
```

###################################################################################################
# IMPORT DATA FROM "ANNOTATION_FILES > LINES" FOLDER
###################################################################################################

### 1. FROM Annotation_files folder,
### 2. USE base R import, to
### 3. GET annotation files and differentially expressed gene data.frames.

```{r fig.keep='all', message=FALSE, warning=FALSE, results='hide',tidy=TRUE, tidy.opts=list(width.cutoff=60)}
#LIST FILES WITHIN "LINES" WORKING DIRECTORY
list.lines<- list.files(pattern=".tabular")
#READ ALL FILES FROM WORKING DIRECTORY USING LIST OF WORKING DIRECTORY
DEG.raw<- as.vector(lapply(list.lines, read.table, header = FALSE, sep = "\t", dec =".", na.strings = "NA", colClasses = c("character",rep("numeric",6))))
DEG.raw<-lapply(DEG.raw, setNames, c("ID", "COUNTS", "FC", "SE", "WALD", "P", "FDR"))
#REMOVE FILENAME EXTENSION
list.lines<-gsub(pattern = ".tabular", replacement ="", x=list.lines)
list.lines<-gsub(pattern = "log2.", replacement ="", x=list.lines)
list.lines<-str_split_fixed(list.lines,pattern = "_",n = 2)[,1]
names(DEG.raw)<-list.lines 
```

###################################################################################################
# PLOT KEGG PATHWAY ENRICHMENT FOR EACH MUTANT FISH LINE USING CLUSTERPROFILER
###################################################################################################

### 1. FROM a list of signficantly differentially expressed genes for each mutant line,
### 2. USE bitr to convert zebrafish ensembl gene ids to entrez gene ids and compareCluster, to
### 3. GET enriched KEGG pathways.

```{r echo=TRUE, fig.keep='all', message=FALSE, warning=FALSE,fig.height = 12,fig.width = 12,tidy=TRUE, tidy.opts=list(width.cutoff=60)}
#FILTER GENE IDS WITH FDR <=0.05
SIG.FDR.FC<-lapply(DEG.raw, function(x)subset(x, FDR <= 0.05, select = c("ID","FC"))) 
#ORDER GENE IDS LOG2 FOLD-CHANGE
SIG.FDR.FC<-lapply(SIG.FDR.FC, function(X)X[order(X$FC,decreasing = T),]) 
#SUBSET TOP 750 UP AND DOWN DIFFERENTIALLY EXPRESSED GENES, RESPECTIVELY
SIG.FDR.FC<-lapply(SIG.FDR.FC, function(X)rbind.data.frame(head(X,750), tail(X,750))) 
#REMOVE EMPTY LISTS
SIG.FDR.FC<-SIG.FDR.FC[lapply(SIG.FDR.FC,nrow)!=0]
#CONVERT ENSEMBL IDS TO ENTREZIDS
DEG.ENTREZ <- lapply(SIG.FDR.FC,function(x)bitr(as.character(unlist(x)), fromType="ENSEMBL", toType=c("ENTREZID","GO"), OrgDb = org.Dr.eg.db)$ENTREZID) 
#KEGG PATHWAY ENRICHMENT
DEG.COMP<- compareCluster(DEG.ENTREZ, fun="enrichKEGG", organism="zebrafish", pvalueCutoff=0.05)
DEG.COMP.filt = DEG.COMP
#ADJUSTED P VALUE FILTER
DEG.COMP.filt@compareClusterResult = DEG.COMP.filt@compareClusterResult[DEG.COMP.filt@compareClusterResult$p.adjust <=0.1,]
#Q VALUE FILTER
DEG.COMP.filt@compareClusterResult = DEG.COMP.filt@compareClusterResult[!is.na(DEG.COMP.filt@compareClusterResult$qvalue),]
#CAPITALISE PATHWAY NAMES
DEG.COMP.filt@compareClusterResult$Description = toupper(DEG.COMP.filt@compareClusterResult$Description) 
#REMOVE METABOLIC PATHWAY FILTER
DEG.COMP.filt@compareClusterResult = DEG.COMP.filt@compareClusterResult[!grepl("METABOLISM|METABOLISM|BIOSYNTHESIS|BIOSYNTHESIS|FATTY|GLYCOLYSIS|GLUCONEOGENESIS|HERPES|FERROPTOSIS|AGE|APELIN|ADRENERGIC|CARDIAC|LECTIN|CARBON|VALINE|PHOTO|PENTOSE|NEUROACTIVE|GNRH|OXIDATIVE|ABC|MUSCLE|OOCYTE|SALMONELLA|MELANOGENESIS|SMOOTH|GAP|NOD-LIKE|TOLL-LIKE|ADIPOCYTOKINE|CALCIUM|ERBB|MTOR|FOCAL|ENDOCYTOSIS|LYSOSOMEGLYCAN|ECM|ACTIN|FANCONI|CAMS|PPAR|PEROXISOMES|PHAGOSOME|TIGHT|RIG-I-LIKE|MITOPHAGY|LYSOSOME|GLYCAN|ADHERENS|FOXO|PEROXISOME|TCA|PHOSPHATIDYL|OTHER|ADHERENS|ANIMAL|VEGF|INSULIN|TGF|LYSINE|SNARE|BASAL|HEDGEHOG|CYTOSOLIC", DEG.COMP.filt@compareClusterResult$Description),]
DEG.COMP.filt@compareClusterResult = DEG.COMP.filt@compareClusterResult[!grepl(".*METABOLISM.*|.*BIOSYNTHESIS*", DEG.COMP.filt@compareClusterResult$Description),]
#ORDER BY GENE NAME
DEG.COMP.filt@compareClusterResult$Cluster<-as.character(DEG.COMP.filt@compareClusterResult$Cluster)
#REORDER ROWS BY PREDEFINING FACTORS LEVELS   
#REMOVE INCOMPLETE DATA
DEG.COMP.filt@compareClusterResult<-DEG.COMP.filt@compareClusterResult[complete.cases(DEG.COMP.filt@compareClusterResult),]
#RETAIN ONLY SHARED PATHWAYS
#ORDER ROWS (PATHWAYS) BY COOCCURENCE OF ENRICHED PATHWAYS
l<-split(DEG.COMP.filt@compareClusterResult$Description, f = DEG.COMP.filt@compareClusterResult$Cluster)
jacc.ind <- lapply(1:length(l),function(x)lapply(1:length(l),function(y)100*length(intersect( as.character(unlist(l[[x]])), as.character(unlist(l[[y]]))))/length(union( as.character(unlist(l[[x]])), as.character(unlist(l[[y]]))))))
jacc.ind<-matrix(unlist(jacc.ind), nrow = length(l), ncol = length(l))
colnames(jacc.ind)<-names(l)
rownames(jacc.ind)<-names(l)
jacc.dist<-dist(jacc.ind, method = "euclidean")
clust.jacc<-hclust(jacc.dist, method="ward.D")
jacc.order<-clust.jacc$labels[clust.jacc$order]
DEG.COMP.filt@compareClusterResult<-DEG.COMP.filt@compareClusterResult[DEG.COMP.filt@compareClusterResult$Cluster %in% jacc.order,]
DEG.COMP.filt@compareClusterResult$Cluster<-factor(DEG.COMP.filt@compareClusterResult$Cluster,levels = unique(jacc.order))
DEG.COMP.filt@compareClusterResult<-DEG.COMP.filt@compareClusterResult[order(DEG.COMP.filt@compareClusterResult$Cluster),]
rownames(DEG.COMP.filt@compareClusterResult)<-NULL
#ORDER COLOUMNS (LINES) BY COOCCURENCE OF ENRICHED PATHWAYS
l.2<-split(f =DEG.COMP.filt@compareClusterResult$Description,  DEG.COMP.filt@compareClusterResult$Cluster)
jacc.ind.2 <- lapply(1:length(l.2),function(x)lapply(1:length(l.2),function(y)100*length(intersect( as.character(unlist(l.2[[x]])), as.character(unlist(l.2[[y]]))))/length(union( as.character(unlist(l.2[[x]])), as.character(unlist(l.2[[y]]))))))
jacc.ind.2<-matrix(unlist(jacc.ind.2), nrow = length(l.2), ncol = length(l.2))
colnames(jacc.ind.2)<-names(l.2)
rownames(jacc.ind.2)<-names(l.2)
jacc.dist.2<-dist(jacc.ind.2, method = "euclidean")
clust.jacc.2<-hclust(jacc.dist.2, method="ward.D2")
jacc.order.2<-clust.jacc.2$labels[clust.jacc.2$order]
DEG.COMP.filt@compareClusterResult<-DEG.COMP.filt@compareClusterResult[DEG.COMP.filt@compareClusterResult$Description %in% jacc.order.2,]
DEG.COMP.filt@compareClusterResult$Description<-factor(DEG.COMP.filt@compareClusterResult$Description,levels = unique(jacc.order.2))
DEG.COMP.filt@compareClusterResult<-DEG.COMP.filt@compareClusterResult[order(DEG.COMP.filt@compareClusterResult$Description),]
rownames(DEG.COMP.filt@compareClusterResult)<-NULL

op <- par(oma=c(5,7,1,1))
par(op)

#PLOT ENRICHED PATHWAY OF EACH LINE FOR BOTH UP AND DOWN REGULATED
p<- dotplot(DEG.COMP.filt, includeAll=T,showCategory = 5, font.size=10, by="rowPercentage")
p + theme(axis.text.x=element_text(angle=90, hjust=0.5, vjust = 0.5))
```

###################################################################################################
# PLOT TOP GENES FROM ENRICHED KEGG PATHWAYS FOR EACH MUTANT FISH LINE USING KEGGREST
###################################################################################################

### 1. FROM a list KEGG pathway IDs enriched for each mutant line,
### 2. USE keggGet to extract pathway-specific IDs, to
### 3. GET a heatmap of top differential expressed genes for each mutant line.

```{r echo=TRUE, fig.keep='all', message=FALSE, warning=FALSE,fig.height = 16,fig.width = 16,tidy=TRUE, tidy.opts=list(width.cutoff=60)}
#KEGG PATHWAY IDS
#DNA
dna<- c("dre03030","dre03430","dre03410","dre03440","dre04115","dre04210")
#CELL CYCLE
cell.cycle<- c("dre04110","dre04218", "dre04010")
#SPLICEOSOME
spliceosome<- c("dre03040","dre03015","dre03013","dre03018")
#ER
er<-c("dre04141","dre03050", "dre03060")
#IMMUNE
immune<-c("mmu04660","mmu05340","dre04330")
#RIBOSOME
ribsome<- c("dre03008","dre03010")
#MATCH EACH PATHWAY WITH A ROW SIDE COLOUR
kegg.codes<-c(dna,cell.cycle,spliceosome,er,immune,ribsome)
kegg.colors<-c(rep("firebrick1",length(dna)),
               rep("deeppink3",length(cell.cycle)),
               rep("cornflowerblue",length(spliceosome)),
               rep("darkorange2",length(er)),
               rep("mediumseagreen",length(immune)),
               rep("gold",length(ribsome))              )
df.keggid.w.color<-data.frame(KEGG.ID=kegg.codes, COLOUR=kegg.colors,stringsAsFactors = F)
#GET GENES FROM EACH KEGG PATHWAY
kegg.pathways<-sapply(kegg.codes,function(X)keggGet(X))
#SUBSET GENE IDS
kegg.ids<-lapply(1:length(kegg.pathways),function(X)kegg.pathways[[X]]$GENE)
#EXTRACT GENE IDS FROM FULL GENE NAMES
kegg.ids<-lapply(kegg.ids,function(X)X[grepl(pattern = ";",X)])
kegg.ids<-lapply(kegg.ids,function(X)str_split_fixed(X,';',2)[,1])
#NAME LIST WITH KEGG PATHWAY IDS
names(kegg.ids)<-kegg.codes
#REPLACE MOUSE GENES WITH ZEBRAFISH ORTHOLOGS FOR T CELL PATHWAYS
mouse.KEGG<-c("mmu04660", "mmu05340")
kegg.ids[names(kegg.ids) %in% mouse.KEGG]<-lapply(kegg.ids[names(kegg.ids) %in% mouse.KEGG],function(X)MOUSE.ORTHOLOGS$GENE[match(X,MOUSE.ORTHOLOGS$M_GENE)])
#ADD KEGG ID COLUMN TO LIST
for (i in 1:length(kegg.ids)){
  kegg.ids[[i]]<-data.frame(GENE=as.character(kegg.ids[[i]]),stringsAsFactors = F)
}
for (i in 1:length(kegg.ids)){
kegg.ids[[i]]<-data.frame(kegg.ids[[i]],KEGG.ID=names(kegg.ids[i]),stringsAsFactors = F)
}
#ROW BIND DATA
df.m.gene.kegg.id<-do.call(rbind, kegg.ids)
rownames(df.m.gene.kegg.id)<-NULL
#JOIN COLORS WITH KEGG.ID
df.m.gene.kegg.id<-left_join(df.m.gene.kegg.id,df.keggid.w.color, by="KEGG.ID")
#REMOVE ROWS WITH MISSING DATA
df.zf<-df.m.gene.kegg.id[complete.cases(df.m.gene.kegg.id),]
#REMOVE DUPLICATED GENES
df.zf<-df.zf[!duplicated(df.zf$GENE),]
#MERGE WITH LOG2.FC DATA FOR EACH LINE
df.zf<-left_join(df.zf,MUTANT.LINE.LOG2.FC, by="GENE")
#REMOVE ROWS WITH MISSING DATA
df.zf<-df.zf[complete.cases(df.zf),]
df.zf$ID<-as.character(df.zf$ID)
#CONVERT TO LONG FORMAT
melt.df.zf<- melt(df.zf,id.vars = colnames(df.zf[1:4]), variable.name = "LINE", value.name = "LOG2.FC",factorsAsStrings = F)
#DEFINE EACH MUTANT BE IT'S ENRICHED PATHWAY
DNA<-c("MCM10", "POLE1", "ATAD5A", "TOP3A", "DNMT1")
CEL<-c("ANAPC1", "NEK7")
SPL<-c("LSM8", "SNAPC3", "GEMIN5","UPF1", "CSTF3")
ER<-c("TBCB", "UBE3D", "UNC45A")
HEM<-c("RAG1", "IKZF1", "FLI1A", "CMYB", "IL7R", "FOXN1", "ZBTB17")
RIB<-c("SPATA5", "NOL9")
#EXTRACT TOP DIFFERENTIALLY EXPRESSED GENES FOR EACH PATHWAY 
#DNA
up.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(DNA,collapse = "|")) & COLOUR == "firebrick1") %>%
    group_by(LINE) %>%
    arrange(-LOG2.FC) %>%  
    top_n(.,2) %>%
    .$ID
down.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(DNA,collapse = "|")) & COLOUR == "firebrick1") %>%
    group_by(LINE) %>%
    arrange(LOG2.FC) %>%  
    top_n(.,-1) %>%
    .$ID
top.IDs<-c(up.DEG,down.DEG)
#SPL
up.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(SPL,collapse = "|")) & COLOUR == "cornflowerblue") %>%
    group_by(LINE) %>%
    arrange(-LOG2.FC) %>%  
    top_n(.,3) %>%
    .$ID
top.IDs<-c(top.IDs,up.DEG,down.DEG)
#CEL
up.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(CEL,collapse = "|")) & COLOUR == "deeppink3") %>%
    group_by(LINE) %>%
    arrange(-LOG2.FC) %>%  
    top_n(.,2) %>%
    .$ID
down.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(CEL,collapse = "|")) & COLOUR == "deeppink3") %>%
    group_by(LINE) %>%
    arrange(LOG2.FC) %>%  
    top_n(.,-1) %>%
    .$ID
top.IDs<-c(top.IDs,up.DEG,down.DEG)
#ER
up.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(ER,collapse = "|")) & COLOUR == "darkorange2") %>%
    group_by(LINE) %>%
    arrange(-LOG2.FC) %>%  
    top_n(.,2) %>%
    .$ID
top.IDs<-c(top.IDs,up.DEG,down.DEG)
#HEM
down.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(HEM,collapse = "|")) & COLOUR == "mediumseagreen") %>%
    group_by(LINE) %>%
    arrange(LOG2.FC) %>%  
    top_n(.,-3) %>%
    .$ID
top.IDs<-c(top.IDs,up.DEG,down.DEG)
#RIB
up.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(RIB,collapse = "|")) & COLOUR == "gold") %>%
    group_by(LINE) %>%
    arrange(-LOG2.FC) %>%  
    top_n(.,2) %>%
    .$ID
down.DEG<-melt.df.zf %>% 
    filter(grepl(LINE,pattern = paste(RIB,collapse = "|")) & COLOUR == "gold") %>%
    group_by(LINE) %>%
    arrange(LOG2.FC) %>%  
    top_n(.,-2) %>%
    .$ID
top.IDs<-unique(c(top.IDs,up.DEG,down.DEG))
#FILTER FOR TOP DEG GENES FROM EACH PATHWAY
top.df.zf<-df.zf[df.zf$ID %in% top.IDs,]
#PERFORM HCLUST ON GENES WITHIN EACH PATHWAY TO SORT DATA FRAME WITHOUT MIXING ROWSIDE COLOURS
#SPLIT DF BY COLOUR INTO SEPARATE LIST
split.df <- split(top.df.zf,top.df.zf$COLOUR)
#REORDER BY MUTANT TABLE ORDER
split.df <-split.df[match(c("mediumseagreen","firebrick1","cornflowerblue", "darkorange2","deeppink3","gold"), names(split.df))]
#ADD IDS TO ROWNAMES
for (x in 1:length(split.df))rownames(split.df[[x]])<-split.df[[x]]$ID
#COMPUTE DISTANCE MATRIX FOR EACH GENE
dists <- lapply(split.df,function(x)dist(x[,5:ncol(x)], method = "euclidean"))
#CONVERT TO HCLUST OBJECT FOR SORTING
h.clust <-lapply(dists,function(x)hclust(x,method = "complete"))
#REORDER DF USING HCLUST OBJECT
figure <- list()
for (i in 1:length(names(split.df))){
  figure[[i]] <- top.df.zf[top.df.zf$COLOUR == names(split.df)[[i]],][h.clust[[i]]$order,]
}
#BIND DATA.FRAME AFTER SORTING
df.order <- unique(do.call(rbind,figure))
df.order$COLOUR<-factor(df.order$COLOUR,levels = c("mediumseagreen", "firebrick1","deeppink3","cornflowerblue","darkorange2","gold"))
df.order<-df.order[order(df.order$COLOUR,decreasing = F),]
#MARGINS
op <- par(oma=c(5,7,1,1))
par(op)
#PLOT HEATMAP
my_palette <- colorRampPalette(c("deeppink","white", "deepskyblue"))(n = 299)
col_breaks = c(seq(-2,-1,length=100), seq(-0.99,0.99,length=100), seq(1,2,length=100))
#REORDER COLUMN INTO PATHWAY GROUPS
mat<-as.matrix(df.order[,5:ncol(df.order)])
mat.col.order<-intersect(NODE_PATHWAY_ANNOTATION$NODE1[order(NODE_PATHWAY_ANNOTATION$ORDER)],colnames(mat)) 
mat<-mat[,match(mat.col.order,colnames(mat))]
#PLOT TOP 5 DIFFERENTIALLY EXPRESSED GENES PER MUTANT LINE
M.sig<-mat
na.index<-sapply(1:ncol(M.sig),function(X)c(as.numeric(rownames(df.order)[head(order(M.sig[,X]),3)]), as.numeric(rownames(df.order)[head(order(M.sig[,X], decreasing = T),3)])))
for (i in 1:ncol(M.sig)){
  M.sig[,i][!rownames (df.order) %in% na.index[,i]]<-NA
  }
M.sig[abs(M.sig)<=0.5]<-NA
M.sig[!is.na(M.sig)]<-"*"
#PLOT
heatmap.2(mat, 
          trace = "none", 
          col = my_palette, 
          breaks = col_breaks, 
          cexRow = 2, 
          cexCol = 2, 
          margin=c(5,20),
          labRow = as.expression(lapply(df.order$GENE, function(a) bquote(italic(.(a))))), 
          labCol = as.expression(lapply(tolower(colnames(mat)), function(a) bquote(italic(.(a))))), 
          RowSideColors = as.character(df.order$COLOUR),
          Rowv = F, 
          dendrogram = "column", 
          key = F,
          cellnote = M.sig, 
          notecol = "black", 
          notecex = 2,
          lwid = c(0.2,5),
          lhei = c(1,12))
```